### PR TITLE
Update Gradient Clipping for TPU Accelerator

### PR DESCRIPTION
**Pull request by kaushikb11**
## What does this PR do?

Add updates to TPU Accelerator:
* Graident Clipping

## Before submitting
- [ ] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [ ] Did you make sure to update the documentation with your changes? (if necessary)
- [ ] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/accelerators/tpu.py [hunk 1]
```diff
@@ -55,3 +56,11 @@ def all_gather(self, tensor: torch.Tensor, group: Optional[Any] = None, sync_gra
         if torch.distributed.is_initialized():
             return xm.all_gather(tensor, group=group, sync_grads=sync_grads)
         return tensor
+
+    def clip_gradients(self, optimizer: Optimizer, grad_clip_val: Union[float, int], norm_type: float = 2.0):
```
Comment (SeanNaren) [inline pytorch_lightning/accelerators/tpu.py:60]: The patched XLA function does not do a check for `grad_clip_val>0` so we might need to include this as done here:

https://github.com/PyTorchLightning/pytorch-lightning/blob/master/pytorch_lightning/plugins/precision/precision_plugin.py#L109-L110

In addition, will this mean we can get rid of the TODO here? https://github.com/PyTorchLightning/pytorch-lightning/blob/master/pytorch_lightning/plugins/precision/precision_plugin.py#L103

The implementation of clip gradients is TPU compliant I think, but I think that's not an issue.

Comment (tchaton) [reply pytorch_lightning/accelerators/tpu.py:60]: Good point !

Comment (kaushikb11) [reply pytorch_lightning/accelerators/tpu.py]: @SeanNaren The issue for `aitextgen`?

The code hangs when it calls xm.save in Transformer’s [save_pretrained](https://github.com/minimaxir/aitextgen/blob/master/aitextgen/train.py#L235) method, when passed xm_save to the save_function param from aitextgen for TPUs.
Code: https://github.com/huggingface/transformers/blob/master/src/transformers/modeling_utils.py#L835

`xm.save` has [rendezvous](https://github.com/pytorch/xla/blob/master/torch_xla/core/xla_model.py#L819) in it. All the core were not syncing up for rendezvous. Hence, the code was hanging.

**Comment by SeanNaren** [comment]
I know tests are annoying for TPUs, but could we add a test to ensure that clipping works? a simple test just to make sure the function is called like the sharded test:

```python
@pytest.mark.parametrize("clip_val", [0, 10])
@RunIf(min_gpus=1, skip_windows=True, amp_native=True, fairscale=True)
@mock.patch('fairscale.optim.oss.OSS.clip_grad_norm')
def test_ddp_sharded_precision_16_clip_gradients(mock_oss_clip_grad_norm, clip_val, tmpdir):
    """
    Ensure that clip gradients is only called if the value is greater than 0.
    """
    model = BoringModel()
    trainer = Trainer(accelerator='ddp_sharded', gpus=1, precision=16, fast_dev_run=True, gradient_clip_val=clip_val)
    trainer.fit(model)
    if clip_val > 0:
        mock_oss_clip_grad_norm.assert_called()
    else:
        mock_oss_clip_grad_norm.assert_not_called()
```

**Review by SeanNaren — APPROVED** [review]

**Review by tchaton — APPROVED** [review]
LGTM !

**Review by awaelchli — APPROVED** [review]
is it compatible with the TPUHalf plugin?
Should we extend clip_gradients there and `raise NotImplementedError`, what do you think?

**Comment by kaushikb11** [comment]
> is it compatible with the TPUHalf plugin?
> Should we extend clip_gradients there and `raise NotImplementedError`, what do you think?

@awaelchli But it could be the default for the TPU accelerator, right?
